File: android/questionparser/QuestionParser.java
package questionparser;

import java.util.Dictionary;
import java.util.Enumeration;
import java.util.LinkedList;
import java.util.List;

public class QuestionParser {
    private static Dictionary<String, Dictionary<String, Integer>> patternDict = QuestionWordGenerator.QUESTION_WORDS;

    public static List<String> phraseSplit(String phrase) {
        String[] words = phrase.split("\\s");
        Dictionary offset = patternDict.get(words[0]);
        int numOffset = 0;
        for(int i=1;i < words.length;i++) {
            if(patternDict.get(words[i]) != null) {
                numOffset = ((int) offset.get(words[i])) + i + 2;
                break;
            }
        }
        String invocationPhrase = "";
        String listPhrase = "";
        for(int i=0;i < words.length;i++) {
            if(i < numOffset) {
                invocationPhrase += words[i] + " ";
            }
            else {
                listPhrase += words[i] + " ";
            }
        }
        invocationPhrase = invocationPhrase.substring(0, invocationPhrase.length() - 1);
        listPhrase = listPhrase.substring(0, listPhrase.length() - 1);
        List<String> splitPhrases = new LinkedList<>();
        splitPhrases.add(invocationPhrase);
        splitPhrases.add(listPhrase);
        return splitPhrases;
    }

    public static boolean questionClassifier(String phrase) {
        String[] words = phrase.split("\\s");
        for(int i=0;i < words.length;i++) {
            if(words[i].equals("or")) {
                Enumeration<String> keys = patternDict.keys();
                for(int j=0; j < words.length;j++) {
                    while(keys.hasMoreElements()) {
                        String key = keys.nextElement();
                        if(key.equals(words[j])) {
                            return true;
                        }
                    }
                }
            }
        }
        return false;
    }
}

File: online_api/api/question_phrase_parser/question_parser.py
from api.question_phrase_parser.question_words_generator import load_question_words

pattern_dict = load_question_words()


def phrase_split(phrase):
    """
    accepts a full phrase string and returns a list of the invocation and list phrase
    :return: returns a list with two strings the invocation phrase and the list phrase
    :rtype: list<str>
    """
    try:
        words = phrase.split()
        offset = pattern_dict[words[0]]
        num_offset = 0
        for i in range(len(words[1:])):
            for key, value in offset.items():
                if words[1:][i] == key:
                    num_offset = value + i + 2
                    break
        invocation_phrase = " ".join(words[:num_offset])
        list_phrase = " ".join(words[num_offset:])
    except Exception:
        invocation_phrase = ""
        list_phrase = phrase
    return [invocation_phrase, list_phrase]


def question_classifier(phrase):
    """
    classifies if a question is a list question
    :param phrase: phrase to test
    :return: if the question is a classifier
    :rtype: bool
    """
    words = phrase.split()
    if 'or' in words:
        if any(x in words for x in pattern_dict.keys()):
            return True
    return False
